packages/legend-query-builder/src/components/watermark/QueryBuilderWatermark.tsx

```diff
@@ -0,0 +1,209 @@
+        },
+        collect: (monitor) => ({
+          isParameterValueDragOver: monitor.isOver({
+            shallow: true,
+          }),
+        }),
+      }),
+      [handleDrop],
+    );
+
+    return (
+      <PanelFormSection>
+        <div className="query-builder__parameter-editor">
+          <PanelDropZone
+            isDragOver={isParameterValueDragOver}
+            dropTargetConnector={dropTargetConnector}
+          >
+            <BasicValueSpecificationEditor
+              valueSpecification={guaranteeNonNullable(watermarkValue)}
+              setValueSpecification={(val: ValueSpecification): void => {
+                watermarkState.setValue(val);
+              }}
+              graph={graph}
+              obseverContext={
+                watermarkState.queryBuilderState.observableContext
+              }
+              typeCheckOption={{
+                expectedType: graph.getPrimitiveType(PRIMITIVE_TYPE.STRING),
+              }}
+              resetValue={(): void => {
+                watermarkState.resetValue();
+              }}
+            />
+          </PanelDropZone>
+        </div>
+      </PanelFormSection>
+    );
+  },
+);
+
+export const QueryBuilderWatermarkEditor = observer(
+  (props: { queryBuilderState: QueryBuilderState }) => {
+    const { queryBuilderState } = props;
+
+    const watermarkState = queryBuilderState.watermarkState;
+
+    const handleClose = (): void => {
+      queryBuilderState.setIsEditingWatermark(false);
+    };
+
+    return (
+      <Dialog
+        open={Boolean(queryBuilderState.isEditingWatermark)}
+        onClose={handleClose}
+        classes={{
+          root: 'editor-modal__root-container',
+          container: 'editor-modal__container',
+          paper: 'editor-modal__content',
+        }}
+      >
+        <Modal darkMode={true} className="editor-modal">
+          <ModalHeader title="Watermark" />
+          <ModalBody>
+            <PanelForm>
+              <PanelFormBooleanField
+                isReadOnly={false}
+                value={watermarkState.value !== undefined}
+                prompt="Add a watermark"
+                update={(): void => watermarkState.toogleWatermark()}
+              />
+              {watermarkState.value && (
+                <>
+                  <WatermarkValueEditor
+                    watermarkState={watermarkState}
+                    watermarkValue={watermarkState.value}
+                  />
+                  <PanelDivider />
+                  <PanelFormListItems title="List of available parameters">
+                    {queryBuilderState.parametersState.parameterStates
+                      .length === 0 && <> No available parameters in query </>}
+                    {queryBuilderState.parametersState.parameterStates
+                      .filter((pState) => {
+                        const isMuliplicityOne =
+                          pState.parameter.multiplicity ===
```

Comment:
I don't think this is a safe check, there's a method to do multiplicity equality, `multiplicityComparator`, let's use that